### PR TITLE
[RISCV] Change vector tuple type's TypeSize to scalable

### PR DESCRIPTION
**Pull request by 4vtomat**
Vector tuple is basically multiple grouped vector, so its size is also
determined by vscale, we need not to model it as a vector type but its
size need to be scalable.

### CODE REVIEW

File: llvm/include/llvm/CodeGen/ValueTypes.td [hunk 1]
```diff
@@ -63,6 +63,7 @@ class VTVecTup<int size, int nf, ValueType dummy_elt, int value>
   let NF = nf;
   let ElementType = dummy_elt;
   let isRISCVVecTuple = true;
+  let isScalable = true;
```
Comment (topperc) [inline llvm/include/llvm/CodeGen/ValueTypes.td:66]: I don't think you should to this. This field is misnamed. It should be called "isScalableVector"

Comment (4vtomat) [reply llvm/include/llvm/CodeGen/ValueTypes.td:66]: Do you mean we should make tuple type a special case and handle it in switch in `getSizeInBits`?

Comment (topperc) [reply llvm/include/llvm/CodeGen/ValueTypes.td:66]: What I don't want is for this to return true for a tuples.

```
    bool isScalableVector() const {                                              
      return (SimpleTy >= MVT::FIRST_SCALABLE_VECTOR_VALUETYPE &&                
              SimpleTy <= MVT::LAST_SCALABLE_VECTOR_VALUETYPE);                  
    }  
```

I think you just need to change `Sc || Ty == aarch64svcount` in this to `Sc || Tup || Ty == aarch64svcount`

```
    TypeSize getSizeInBits() const {
      static constexpr TypeSize SizeTable[] = {
#define GET_VT_ATTR(Ty, N, Sz, Any, Int, FP, Vec, Sc, Tup, NF, NElem, EltTy) \
    TypeSize(Sz, Sc || Ty == aarch64svcount /* FIXME: Not in the td. */),
#include "llvm/CodeGen/GenVT.inc"                                                
#undef GET_VT_ATTR
```

Comment (4vtomat) [reply llvm/include/llvm/CodeGen/ValueTypes.td:66]: > What I don't want is for this to return true for a tuples.
> 
> ```
>     bool isScalableVector() const {                                              
>       return (SimpleTy >= MVT::FIRST_SCALABLE_VECTOR_VALUETYPE &&                
>               SimpleTy <= MVT::LAST_SCALABLE_VECTOR_VALUETYPE);                  
>     }  
> ```
> 
I see

> I think you just need to change `Sc || Ty == aarch64svcount` in this to `Sc || Tup || Ty == aarch64svcount`
> 
> ```
>     TypeSize getSizeInBits() const {
>       static constexpr TypeSize SizeTable[] = {
> #define GET_VT_ATTR(Ty, N, Sz, Any, Int, FP, Vec, Sc, Tup, NF, NElem, EltTy) \
>     TypeSize(Sz, Sc || Ty == aarch64svcount /* FIXME: Not in the td. */),
> #include "llvm/CodeGen/GenVT.inc"                                                
> #undef GET_VT_ATTR
> ```
Yeah, it's easier lol~

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -21309,6 +21310,25 @@ bool RISCVTargetLowering::splitValueIntoRegisterParts(
     return true;
   }
 
+  if (ValueVT.isRISCVVectorTuple() && PartVT.isRISCVVectorTuple()) {
+    unsigned ValNF = ValueVT.getRISCVVectorTupleNumFields();
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21314]: Is this only called when PartVT and ValueVT are the same type? That shouldn't require any node to be created.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 3]
```diff
@@ -21378,6 +21382,9 @@ SDValue RISCVTargetLowering::joinRegisterPartsIntoValue(
     return Val;
   }
 
+  if (ValueVT.isRISCVVectorTuple())
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21385]: Why do we need this now?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 2 above)
Comment (4vtomat) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21314]: I see, let me have another patch to fix it, thanks!

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 3 above)
Comment (4vtomat) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21385]: Actually we can remove it after removing tuple from scalable vector type.

File: llvm/lib/CodeGenTypes/LowLevelType.cpp [hunk 4]
```diff
@@ -22,6 +22,11 @@ LLT::LLT(MVT VT) {
     init(/*IsPointer=*/false, asVector, /*IsScalar=*/!asVector,
          VT.getVectorElementCount(), VT.getVectorElementType().getSizeInBits(),
          /*AddressSpace=*/0);
+  } else if (VT.isRISCVVectorTuple()) {
+    // TODO: Correctly model RISC-V vector tuple type
```
Comment (wangpc-pp) [inline llvm/lib/CodeGenTypes/LowLevelType.cpp:26]: Should we include this change? We don't have LLT support of tuples before, whether it is scalable or not.

Comment (4vtomat) [reply llvm/lib/CodeGenTypes/LowLevelType.cpp:26]: Current approach makes it use scalar construction, however once we change the TypeSize, `VT.getSizeInBits()` will no longer convertible directly to `size_t`.

**Comment by 4vtomat** [comment]
Ping~

File: llvm/lib/CodeGenTypes/LowLevelType.cpp (hunk 4 above)
Comment (topperc) [reply llvm/lib/CodeGenTypes/LowLevelType.cpp:26]: Setting IsScalar = true seems wrong thought it was the old behavior. What if we use the final else that sets all fields to `false` or `0`?

File: llvm/include/llvm/CodeGen/ValueTypes.h [hunk 5]
```diff
@@ -185,7 +185,8 @@ namespace llvm {
 
     /// Return true if the type is a scalable type.
     bool isScalableVT() const {
-      return isScalableVector() || isScalableTargetExtVT();
+      return isScalableVector() || isScalableTargetExtVT() ||
+             isRISCVVectorTuple();
```
Comment (topperc) [inline llvm/include/llvm/CodeGen/ValueTypes.h:189]: Should isRISCVVectorTuple() be part of isScalableTargetExtVT()?

Comment (4vtomat) [reply llvm/include/llvm/CodeGen/ValueTypes.h:189]: Yeah, it should

File: llvm/lib/CodeGenTypes/LowLevelType.cpp (hunk 4 above)
Comment (topperc) [reply llvm/lib/CodeGenTypes/LowLevelType.cpp:26]: A tuple is not a scalar. Why is setting IsScalar correct?

Comment (4vtomat) [reply llvm/lib/CodeGenTypes/LowLevelType.cpp]: I think after adding tuple type to `isScalableTargetExtVT`, we no longer need this lol

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 2 above)
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21314]: Add `#ifndef NDEBUG`?

File: llvm/include/llvm/CodeGenTypes/MachineValueType.h [hunk 6]
```diff
@@ -306,12 +306,14 @@ namespace llvm {
     /// be set and the runtime size will be a positive integer multiple of the
     /// base size.
     TypeSize getSizeInBits() const {
+      // clang-format off
```
[the remark this answers is not in the document]
Comment (4vtomat) [reply llvm/include/llvm/CodeGenTypes/MachineValueType.h:309]: Good idea!

**Review by wangpc-pp — APPROVED** [review]
LGTM.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**